File: clojure/largest-series-product/src/largest_series_product.clj
(ns largest-series-product)

(defn largest-product [n digits]
  (when-not (re-matches #"\d*" digits)
    (throw (IllegalArgumentException. "Invalid digit")))
  (when-not (<= n (count digits))
    (throw (IllegalArgumentException. "Span must be <= number of digits")))
  (if (zero? n)
    1
    (->> digits
         (map #(Character/digit ^Character % 10))
         (partition n 1)
         (map (partial reduce *))
         (apply max))))

File: scala/largest-series-product/src/main/scala/LargestSeriesProduct.scala
object Series:
  def largestProduct(span: Int, digits: String): Option[Int] =
    if span < 0 || span > digits.length || digits.exists(!_.isDigit) then None
    else if span == 0 then Some(1)
    else Some(digits.sliding(span).map(_.map(_.asDigit).product).max)
